[Documentation][feature/otx] Add Object Detection Tutorial (#1469)
open-edge-platform/training_extensions: docs/source/tutorials/otx_cli/how_to_train/detection.rst

Object Detection model
======================

This tutorial reveals end-to-end solution from installation to model deploying for object detection task on a certain example.

> On this page we show how to train, validate, export and optimize ATSS model on BCCD public dataset.

Reviewer: wgist

> To learn how to deploy the trained model, refer to: :doc:`../deploy`.

Reviewer: will we separate deploy and demo tutorials? Can we combine them? I think there is no so much info to separate them

Author: We don't have much materials for them, it's true. But IMO it's good to have small tutorials for atomic operations, don't you think?

Reviewer: Okay, let's see how it will look like. We can combine them later if needed

> To learn, how to run the demo, refer to the demo tutorial [#TODO link].

Reviewer 2: Could you elaborate that we have 3 templates and what are the differences (small, medium, large, etc) between these 3 templates? Then, we can say what model we will be using in this tutorial. It'd be nice if you could add ATSS paper reference link. Probably our paper as well.

Author: We wanted to move all explanations about templates characteristics and their description to the `Explanation section` (which is under development right now). We'll add paper links there as well (including our great paper :)).

To learn how to run the demo and visualize results, refer to: :doc:`../demo`.

The process has been tested on the following configuration.

- Ubuntu 20.04
- NVIDIA GeForce RTX 3090
- CUDA Toolkit 11.4 

*************************
Setup virtual environment
*************************

> You can follow the installation process from a :doc:`quick_start quide <../../../get_started/quick_start>` to create a universal virtual environment for all tasks. On the other hand, you can create a task-specific environment following the process below.

Reviewer 3: ```suggestion
You can follow the installation process from a :doc:`quick_start guide <../../../get_started/quick_start>` to create a universal virtual environment for all tasks. On the other hand, you can create a task-specific environment following the process below.
```

> You can follow the installation process from a quick_start quide [#TODO link] to create a univesal virtual environment for all tasks. On other hand, to save memory and time, you can create task-specific environment following the process below.

Reviewer 2: On "the" other hand ~~, to save memory and time,~~ you can create "a" task-specific environment following the process below.

Author: It's a good point, actually. I was sure, what we need less depencies for detection environment, so we're saving memory and time, installing task-specific environment. But turned up, that detection enviroment uses even more memory than virtualenv made from `otx/algorithms/init_venv.sh`. So I'll update it

1. Install prerequisites with:

.. code-block::

    sudo apt-get install python3-pip python3-venv
    # verify your python version
    python3 --version; pip3 --version; 

    Python 3.8.10

>     pip 20.0.2 from /usr/lib/python3/dist-packages/pip (python 3.8)

Reviewer 3: I'm not sure this step is really needed.
Without sudo api install, `pip` or `python -m pip` would work.
Could you fix quick start guide and this part if it's the case?

2. Create and activate a virtual environment for the obect detection task.
The following example creates a virtual environment in the ``det_venv`` folder for detection task.

.. code-block::

    # create virtual env
    bash ./otx/algorithms/detection/init_venv.sh det_venv
    # activate virtual env

>     source det_venv/bin/activate

Reviewer 3: This part will be changed soon as like:

- Virtual env
- Install prerequisite (pytorch,...)
- `pip install -e .` or `pip install otx`

But let's leave this as is for now. :)

>     bash ./otx/algorithms/detection/init_venv.sh det_venv

Reviewer 2: Does this really work? We will remove external, ote_sdk, ote_cli soon. Users should not need to install with this `init_venv.sh` in external as submodule is outdated in feature/otx, instead, users should install with `otx/algorithms/init_venv.sh`.

Author: Yes, it works, but it requires a few more packages to be added in detection requirements (I'll update it in a separate PR along with code changes). This `init_venv.sh` is not in external or submodule. It's the same init_venv as `otx/algorithms/init_venv.sh`, that also installs packages for detection from `requirements/detection.txt`

***************************
Dataset preparation
***************************

> 1. Download a public `BCCD dataset <https://public.roboflow.com/object-detection/bccd/3>`_ (login required). Log in, click ``Download`` button and chose ``Terminal`` option. You will get the code line like this, but with your personal API key.

Reviewer 2: Is this really necessary? There are a few datasets available for detection stored under `data/`. Users should not have to go to external sites, register and download datasets just to try it out, and we also don't maintain BCCD so we can't fix it if it's corrupted in the future. I suggest we use the annotated detection data in `data/airport`.

Author: Yes, actually I agree with you that it'd be much more convenient to train the model on the same data that we already have in the repository. But when we discussed the tutorial, there was an argument that we need to train a real network and get real quality for users, so our toy dataset is not applicable for this purpose.
But I agree that it'd be really convenient not to rely on 3rd party repository. 
@kprokofi What do you think?

Reviewer: Yeah, the main point why we use BCCD dataset that we need to get feasible results in the end. `data/airport` is used for testing purposes and don't give any sufficient results as it contains only a few images. The good point that we don't maintain BCCD. We return again to the issue, that we need some storage to store datasets. For explanation section, for example, we will need maintain 3-4 datasets per  task - it is also a good way to track the quality of our pipeline.

Reviewer: Maybe we can find a small dataset without a registration for now to add direct link, can't we?

Author: Actually, yes. I found out, that one of the small datasets we used now have the COCO annotation and is reachable via link. I'll replace the dataset. Thank you for conversation

Reviewer 2: @kprokofi @GalyaZalesskaya It's nice to generate results on a real-life dataset but this notebook is more like a demo and the flow feels clunky having to go Roboflow register and all that.

That said, I don't mind us using BBCD and even better to host it, however, do we need to acquire a license for it and inform the creators? The reason I propose to use the dataset in repo is that they are created by ourselves, thus, we won't get into copyright disputes.

Author: @eugene123tw @kprokofi As I [can see](https://public.roboflow.com/object-detection/bccd), BBCD dataset is released under the MIT license, which gives us rights to distribute it, copy and modify it.
![image](https://user-images.githubusercontent.com/33455576/211545677-56c6b69e-36cf-4168-87d0-2071c3ab5574.png).
So we need to find, where to store it, and will be able to provide the direct link :) 
As for now, we can keep WGISD dataset, I think

.. code-block::

  curl -L "https://public.roboflow.com/ds/<YOUR_API_KEY>" > bccd.zip;
  unzip bccd.zip; rm bccd.zip

#TODO insert data Sample

2. Check the file structure of downloaded dataset, which should be as following.

.. code-block::

  BCCD
  ├── README.dataset.txt
  ├── README.roboflow.txt
  ├── test/
      ├── _annotations.coco.json
      └── <images>
  ├──train/
      ├──_annotations.coco.json
      └── <images>
  └──valid/
      ├──_annotations.coco.json
      └── <images>

> 3. ``(Optional)`` To simplify the comand line functions calling, you may create a ``data.yaml`` file with annotations info and pass it as a parameter. The content of the ``training_extesions/data.yaml`` for BBCD dataset should have absolete paths and will be similar to that:

Reviewer 2: Could you move this part to the 3rd step of `Training`? It seems odd as this file is passed to training argparse, and should belong in `otx train` section

Author: I think, it's a part of dataset preparation for further actions. Also, it's used by `eval`, `optimize` commands, so maybe `train` section is not the best place for that.

.. code-block::

  {'data': 
    {'train': 
      {'ann-files': '/home/gzalessk/training_extensions/BBCD/train/_annotations.coco.json',

>        'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/train'},

Reviewer 2: I think explaining what data-roots is helpful. Sth like `data-roots` is the folder storing images. We should change `data-roots` from plural to singular `data-root`.

Author: Good point, I'll add some explanation

    'val':
      {'ann-files': '/home/gzalessk/training_extensions/datasets/BBCD/valid/_annotations.coco.json',
       'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/valid'},
    'test':
      {'ann-files': '/home/gzalessk/training_extensions/datasets/BBCD/test/_annotations.coco.json',
       'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/test'}
    }
  }

*********
Training
*********

> 1. Before training you need to chose, which object detection model will you use. The list of supported templates for object detection is available with the command line below. 

Reviewer 2: Before training, you need to "choose" which object detection model you will use. The list of supported templates for object detection is available with the command line below.

  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+
  |    TASK   |                 ID                |  NAME |                                    PATH                                   |
  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+
  | DETECTION |   Custom_Object_Detection_YOLOX   | YOLOX | otx/algorithms/detection/configs/detection/cspdarknet_yolox/template.yaml |
  | DETECTION |  Custom_Object_Detection_Gen3_SSD |  SSD  |  otx/algorithms/detection/configs/detection/mobilenetv2_ssd/template.yaml |
  | DETECTION | Custom_Object_Detection_Gen3_ATSS |  ATSS | otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml |
  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+

2. ``otx train`` trains a model (a particular model template) on a dataset and saves results in two files:

- weights.pth - a model snapshot
- label_schema.json - a label schema used in training, created from a dataset

> These files can be used by other commands: ``export``, ``eval``, ``deploy`` and ``demo``.

Reviewer 2: I don't think we need this sentence as we will be explaining each command in the following section.

Author: Here I wanted to specify, what training part cannot be skipped and the user cannot evaluate just a PyTorch weights from other source. Should I formulate it more clearly?

  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+
  | DETECTION |   Custom_Object_Detection_YOLOX   | YOLOX | otx/algorithms/detection/configs/detection/cspdarknet_yolox/template.yaml |
  | DETECTION |  Custom_Object_Detection_Gen3_SSD |  SSD  |  otx/algorithms/detection/configs/detection/mobilenetv2_ssd/template.yaml |
  | DETECTION | Custom_Object_Detection_Gen3_ATSS |  ATSS | otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml |
  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+

> 2. ``otx train`` trains a model (a particular model template) on a dataset and results in two files:

Reviewer 3: Nit: for numbering, we can use just `1.` for all list item, so that we don't need to care about add, delete, reorder.

- ``weights.pth`` - a model snapshot
- ``label_schema.json`` - a label schema used in training, created from a dataset

These are needed as inputs for the further commands: ``export``, ``eval``,  ``optimize``,  ``deploy`` and ``demo``.

3. To have a specific example in this tutorial, all commands will be run on the ATSS model. It's a medium model, that achieves relatively high accuracy, while keeping the inference fast.

The following command line starts 1 GPU training of the medium object detection model on BCCD dataset:

.. code-block::

  (detection) ...$ otx train otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --train-ann-files data/wgisd/coco_annotations/train_bbox_instances.json 
                            --train-data-roots  data/wgisd/data
                            --val-ann-files data/wgisd/coco_annotations/test_bbox_instances.json 
                            --val-data-roots data/wgisd/data 
                            --save-model-to outputs

>                             --work-dir outputs/logs

Reviewer 3: Do we need both? (sorry, I'm not sure)
@harimkang Could you advise on this?

Author: `save-model-to` saves the final snapshot and label_schema. While `work-dir` saves all training logs and files for tensorbord, what can be useful for a deep research. I'll explain more about it in tutorial, thank you

                            --gpus 1

If you created ``data.yaml`` file in previous step, you can simplify the training by passing it in ``--data`` parameter:

.. code-block::

  (detection) ...$ otx train otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml
                            --save-model-to outputs
                            --work-dir outputs/logs

>                             --gpus 1

Reviewer 3: This might mean GPU ID 1, not using 1 GPU. @eunwoosh Could you confirm?

Author: You're right, will fix that

  ...
  ---------------iou_thr: 0.5---------------

  +-----------+-----+------+--------+-------+
  | class     | gts | dets | recall | ap    |
  +-----------+-----+------+--------+-------+
  | Platelets | 76  | 310  | 1.000  | 0.897 |
  | RBC       | 819 | 4070 | 0.994  | 0.903 |
  | WBC       | 72  | 516  | 1.000  | 0.988 |
  +-----------+-----+------+--------+-------+
  | mAP       |     |      |        | 0.929 |
  +-----------+-----+------+--------+-------+
  2022-12-29 01:08:52,434 | INFO : run task done.
  2022-12-29 01:08:53,010 | INFO : Adjusting the confidence threshold
  2022-12-29 01:08:53,520 | INFO : Setting confidence threshold to 0.32500000000000007 based on results
  2022-12-29 01:08:53,521 | INFO : Final model performance: Performance(score: 0.8315842078960519, dashboard: (17 metric groups))

***********
Validation
***********

1. ``otx eval`` runs evaluation of a trained model on a particular dataset.

Eval function receives test annotation information and model snapshot, trained in previous step.

> Please note, that ``label schema.json`` file should be located in the same folder with model snaphot, as it contains meta information about the dataset .

Reviewer 2: ```suggestion
Please note, ``label_schema.json`` file contains meta-information about the dataset and it should be located in the same folder as the model snapshot.
```

The default metric measured is F1 measure.

> 2. The command below evaluates snaphot in ``outputs`` folder on BCCD dataset and saves results to ``outputs/performance`` :

Reviewer 2: ```suggestion
2. The command below evaluates snapshot in ``outputs`` folder on BCCD dataset and saves results to ``outputs/performance`` :
```

> 2. That's how we can evaluate the snapshot in ``outputs`` folder on BCCD dataset and save results to ``outputs/performance``:

Reviewer: wgisd (please check everywhere)

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --test-ann-files BBCD/valid/_annotations.coco.json 
                            --test-data-roots  BBCD/valid 
                            --load-weights outputs/weights.pth
                            --save-performance outputs/performance.json
  

> If ``data.yaml`` was created, the command can be simplified by passing it for a ``--data`` parameter. Note, that this line will run validation on the test set (not validation set):

Reviewer 2: ```suggestion
If ``data.yaml`` was created, the command can be simplified by passing it for a ``--data`` parameter. Note, with ``data.yaml`` , it runs evaluation on test JSON annotation file (not validation JSON annotation file). 
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml 
                            --load-weights outputs/weights.pth
                            --save-performance outputs/performance.json

> The validation output will look as following:

Reviewer 2: ```suggestion
The evaluation output will look as follows:
```

>                             --save-performance outputs/performance.json

Reviewer 3: If we have `data.yaml` we don't even need to specify in the CLI args.
@sungmanc Could you confirm?

Author: Yes, you're right, but I wanted for the tutorial to show the passing arguments explicitly for a better understanding

.. code-block::

  2022-12-29 01:32:00,505 | INFO : run task done.
  2022-12-29 01:32:01,215 | INFO : Inference completed
  2022-12-29 01:32:01,216 | INFO : called evaluate()
  2022-12-29 01:32:01,527 | INFO : F-measure after evaluation: 0.8315842078960519

> Additionally, you can tune testing parameters such as confidence threshold via comand line. You can read more about template-specific parameters for validation in quick start [#TODO link].

Reviewer 2: ```suggestion
Additionally, you can tune evaluation parameters such as confidence threshold via the command line. You can read more about template-specific parameters for validation in quick start [#TODO link].
```

> For example, to increase the confidence treshold to decrease the number of False Positive predictions (there you have prediction, but don't have annotated object for it) update the evaluation comand line as it's shown below. 

Reviewer 2: ```suggestion
For example, if there are too many False-Positive predictions, you can suppress the number of False-Positives by increasing the confidence threshold as follows:
```

> Please note, that by default confidence treshold is detected automatically based on result to maximize final F1 metric. So, to set custom confidence trashold, please disable ``result_based_confidence_threshold`` option.

Reviewer 2: ```suggestion
Please note, by default, the optimal confidence threshold is detected based on validation results to maximize the final F1 metric. So, to set a custom confidence threshold, please disable ``result_based_confidence_threshold`` option.
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --data data.yaml 
                            --load-weights outputs/weights.pth
                            params 
                            --postprocessing.confidence_threshold 0.5
                            --postprocessing.result_based_confidence_threshold false 

...

> 2023-01-03 18:55:01,956 | INFO : F-measure after evaluation: 0.6274238227146813

Reviewer 2: ![Screenshot from 2023-01-06 10-06-04](https://user-images.githubusercontent.com/13793254/210978662-6eab2fa9-67bf-4014-a667-b831a5dcd0be.png)

The format looks wrong.

3. The output of ``./outputs/performance.json`` consists of dict with target metric name and its value.

.. code-block::

  {"f-measure": 0.8315842078960519}

*********
Export
*********
1. ``otx export`` exports a trained Pytorch `.pth` model to the OpenVINO™ Intermediate Representation (IR) format in order to efficiently run it on Intel hardware. Also, the resulting IR model is required to run POT optimization in the section below.

2. The command below performs exporting of the trained model ``outputs/weights.pth`` in previous section and saves the exported model to the ``outputs/openvino/`` folder.

.. code-block::

  (detection) ...$ otx export otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                              --load-weights outputs/weights.pth
                              --save-model-to outputs/openvino/

  ...

  2022-12-29 01:39:11,980 | INFO : Exporting completed
  2022-12-29 01:39:11,980 | INFO : run task done.
  2022-12-29 01:39:11,990 | INFO : Exporting completed

> 3. You can check the accuracy of exported model as simple as accuracy of the ``.pth`` model, using ``otx eval`` with the path of IR model.

Reviewer 2: ```suggestion
3. You can check the accuracy of the IR model and the consistency between the exported model and the PyTorch model, using ``otx eval`` with the IR model path.
```

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --test-ann-files BBCD/valid/_annotations.coco.json 
                            --test-data-roots  BBCD/valid 
                            --load-weights outputs/openvino/openvino.xml
                            --save-performance outputs/openvino/performance.json
  
  ...

*************
Optimization
*************

1. ``otx optimize`` optimizes a model using NNCF or POT depending on the model format.

- NNCF optimization is used for trained snapshots in a framework-specific format such as checkpoint (pth) file from Pytorch. It starts training-aware quantization based on the obtained weights from the training stage.
- POT optimization is used for models exported in the OpenVINO™ IR format. It decreases floating-point precision to integer precision of the exported model by performing the post-training optimization.

> The function results with a following files, which could be used to run ``otx demo``[link]:

Reviewer 2: ```suggestion
The function results with the following files, which could be used to run ``otx demo``[link]:
```

Object Detection model
======================

#TODO: Made Table of Content for this page?

> This tutorial reveals step by step procedure, how to install OTE CLI and train Object detection model on BBCD public dataset.

Reviewer: I would suggest to give more end-to-end understanding for user, like:

"This tutorial reveal end-to-end solution from installation to model deploying for object detection task. We show how to train, validate, optimize and deploy ATSS model on BBCD public dataset" - smth like that

The process has been tested on the following configuration.

- Ubuntu 20.04
- NVIDIA GeForce RTX 3090
- CUDA Toolkit 11.4 

************************************

> Setup OpenVINO™ Training Extensions

Reviewer: It has been done in quick start section (full installation). I think we need to present code lines for OD task only (partly installation for specific task) or just give a reference link. What do you think?

Author: I think the tutorial should have information to do end-to-end training, so it could be used independently from quick start. But you're right, the cloning part is redundant here, I'll move it to a quick start. And a link will also be useful

************************************

1. Clone the training_extensions repository with the following commands:

.. code-block::

    git clone https://github.com/openvinotoolkit/training_extensions.git
    cd training_extensions
    git checkout develop

2. Install prerequisites with:

.. code-block::

    sudo apt-get install python3-pip python3-venv
    # verify your python version
    python3 --version; pip3 --version; 

Output should be similar to that.

.. code-block::
  
    Python 3.8.10
    pip 20.0.2 from /usr/lib/python3/dist-packages/pip (python 3.8)

3. Create and activate a virtual environment for the obect detection task, then install the ote_cli.
The following example shows that creating virtual environment to the ``det_venv`` folder in your current directory for detection task.

.. code-block::

    # create virtual env
    bash ./otx/algorithms/detection/init_venv.sh det_venv
    # activate virtual env
    source det_venv/bin/activate

***************************
Dataset preparation
***************************

> 1. Download a public `BCCD dataset <https://public.roboflow.com/object-detection/bccd/3>`_ (login required). Log in, put ``Download`` button and chose ``Terminal`` option. You will get the code line like this, but with your personal API key.

Reviewer: ```suggestion
1. Download a public `BCCD dataset <https://public.roboflow.com/object-detection/bccd/3>`_ (login required). Log in, click ``Download`` button and chose ``Terminal`` option. You will get the code line like this, but with your personal API key.
```

.. code-block::

  curl -L "https://public.roboflow.com/ds/<YOUR_API_KEY>" > bccd.zip;
  unzip bccd.zip; rm bccd.zip

#TODO insert data Sample

2. Check the file structure of downloaded dataset, which should be as following.

.. code-block::

  BCCD
  ├── README.dataset.txt
  ├── README.roboflow.txt
  ├── test/
      ├── _annotations.coco.json
      └── <images>
  ├──train/
      ├──_annotations.coco.json
      └── <images>
  └──valid/
      ├──_annotations.coco.json
      └── <images>

> 3. ``(Optional)`` To simplify the comand line functions calling, you may create a ``data.yaml`` file with annotations info and pass it as a parameter. The content of the ``training_extesions/data.yaml`` for BBCD dataset should have absolete paths and will be similar to that:

Reviewer: How to use "data.yaml" after its creation? I don't see an example of its usage. Either we delete it at all and use only "--anotations" or give an example how to use this functionality

.. code-block::

  {'data': 
    {'train': 

>       {'ann-files': '/home/gzalessk/training_extensions/BBCD/train/_annotations.coco.json',

Reviewer: please avoid your home paths

       'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/train'},
    'val':
      {'ann-files': '/home/gzalessk/training_extensions/datasets/BBCD/valid/_annotations.coco.json',
       'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/valid'},
    'test':
      {'ann-files': '/home/gzalessk/training_extensions/datasets/BBCD/test/_annotations.coco.json',
       'data-roots': '/home/gzalessk/training_extensions/datasets/BBCD/test'}
    }
  }

*********
Training
*********

1. Before training you need to chose, which object detection model will you use. The list of supported templates for object detection is available with the command line below. 

.. note::

  The characteristics and detailed comparison of the models could be found in Explanation section [#TODO link].

>   To modify the arhitucture of supported models with various backbones check the Advanced tutorial for dataset modification [#TODO link].

Reviewer: ```suggestion
  To modify the architecture of supported models with various backbones, please refer to the advanced tutorial for model customization  [#TODO link].
```

  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+
  | DETECTION |   Custom_Object_Detection_YOLOX   | YOLOX | otx/algorithms/detection/configs/detection/cspdarknet_yolox/template.yaml |
  | DETECTION |  Custom_Object_Detection_Gen3_SSD |  SSD  |  otx/algorithms/detection/configs/detection/mobilenetv2_ssd/template.yaml |
  | DETECTION | Custom_Object_Detection_Gen3_ATSS |  ATSS | otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml |
  +-----------+-----------------------------------+-------+---------------------------------------------------------------------------+

2. ``otx train`` trains a model (a particular model template) on a dataset and saves results in two files:

- weights.pth - a model snapshot
- label_schema.json - a label schema used in training, created from a dataset

These files can be used by other commands: ``export``, ``eval``, ``deploy`` and ``demo``.

============================
Template-specific parameters
============================

> In order to tune training parameters such as batch size, learning rate, a various set of parameters can be updated via comand line.

Reviewer: We already have the general description of this in the quick start. Let's be more concrete here and give a specific example. (lr=0.005, bs=8, warmup_iters, etc.)

                          header: Number of iterations for learning rate warmup
                          type: INTEGER
                          default_value: 3
                          max_value: 10000
                          min_value: 0

    ...

3. For tutorial purposes, all examples will be run on the ATSS model. This comand line starts 1 GPU training of the medium object detection model on BCCD dataset.

.. code-block::

  (detection) ...$ otx train otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --train-ann-files BBCD/train/_annotations.coco.json 
                            --train-data-roots  BBCD/train 
                            --val-ann-files BBCD/valid/_annotations.coco.json 
                            --val-data-roots BBCD/valid 
                            --save-model-to outputs
                            --work-dir outputs/logs
                            --gpus 1

To decrease batsch size or tune other trainig parameters, extend the comand line above with the following line.

.. code-block::

                            params --learning_parameters.batch_size 4 ...

> The result of the training are weights.pth and label_schema.json, located in ``outputs`` folder, and logs in the ``outputs/logs`` dir.

Reviewer: ```suggestion
The result of the training are ``weights.pth`` and ``label_schema.json``, located in ``outputs`` folder, and logs in the ``outputs/logs`` dir.
```

  2022-12-29 01:00:30,180 | INFO : Epoch [1][1/32]        lr: 1.333e-03, eta: 282 days, 22:46:42, time: 38.204, data_time: 0.462, memory: 4669, current_iters: 0, loss_cls: 1.1113, loss_bbox: 0.5567, loss_centerness: 0.5920, loss: 2.2600, grad_norm: 3.6441

  ...
  ---------------iou_thr: 0.5---------------

  +-----------+-----+------+--------+-------+
  | class     | gts | dets | recall | ap    |
  +-----------+-----+------+--------+-------+
  | Platelets | 76  | 310  | 1.000  | 0.897 |
  | RBC       | 819 | 4070 | 0.994  | 0.903 |
  | WBC       | 72  | 516  | 1.000  | 0.988 |
  +-----------+-----+------+--------+-------+
  | mAP       |     |      |        | 0.929 |
  +-----------+-----+------+--------+-------+
  2022-12-29 01:08:52,434 | INFO : run task done.
  2022-12-29 01:08:53,010 | INFO : Adjusting the confidence threshold
  2022-12-29 01:08:53,520 | INFO : Setting confidence threshold to 0.32500000000000007 based on results
  2022-12-29 01:08:53,521 | INFO : Final model performance: Performance(score: 0.8315842078960519, dashboard: (17 metric groups))

***********
Validation
***********

1. ``otx eval`` runs evaluation of a trained model on a particular dataset.

> Eval function receives test annotation information and folder that contains a model snapshot and label schema.

Reviewer: not folder, but weights direct path

> The default metric measured is f1.

Reviewer: Do we have a functionality to change default metric?

Author: Not for now, according to [this line](https://github.com/openvinotoolkit/training_extensions/blob/c076902a791e069cb89abab89550cc13449d6b53/otx/algorithms/detection/tasks/inference.py#L202), even if it's passed, it's ignored

In order to tune testing parameters such as confidence threshold, a various set of parameters can be updated via comand line.

.. code-block:: 

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml params --help
  usage: otx eval template params [-h] [--postprocessing.confidence_threshold CONFIDENCE_THRESHOLD] [--postprocessing.result_based_confidence_threshold RESULT_BASED_CONFIDENCE_THRESHOLD]
                                  [--nncf_optimization.enable_quantization ENABLE_QUANTIZATION] [--nncf_optimization.enable_pruning ENABLE_PRUNING]

  optional arguments:
    -h, --help            show this help message and exit
    --postprocessing.confidence_threshold CONFIDENCE_THRESHOLD
                          header: Confidence threshold
                          type: FLOAT
                          default_value: 0.35
                          max_value: 1
                          min_value: 0

>     --postprocessing.result_based_confidence_threshold RESULT_BASED_CONFIDENCE_THRESHOLD

Reviewer: It is not clear what is it. Maybe write additional description? 
Also, maybe we add confidence threshold in the below command? (just for example)

                          header: Result based confidence threshold
                          type: BOOLEAN
                          default_value: True
    ...

2. The command below evaluates snaphot in ``outputs`` folder on BCCD dataset and saves results to ``outputs/performance`` :

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --test-ann-files BBCD/valid/_annotations.coco.json 
                            --test-data-roots  BBCD/valid 
                            --load-weights outputs/weights.pth
                            --save-performance outputs/performance.json
  
  ...

  2022-12-29 01:32:00,505 | INFO : run task done.
  2022-12-29 01:32:01,215 | INFO : Inference completed
  2022-12-29 01:32:01,216 | INFO : called evaluate()
  2022-12-29 01:32:01,527 | INFO : F-measure after evaluation: 0.8315842078960519

3. The output of ``./outputs/performance.json`` consists of dict with target metric name and its value.

.. code-block::

  {"f-measure": 0.8315842078960519}

*********
Export
*********

> 1. ``otx export`` exports a trained pth model to the OpenVINO format in order to efficiently run it on Intel hardware. Also, the resulting IR model is required to run POT optimization in section below.

Reviewer: ```suggestion
1. ``otx export`` exports a trained Pytorch `.pth` model to the OpenVINO™ intermediate representation format in order to efficiently run it on Intel hardware. Also, the resulting IR model is required to run POT optimization in the section below.
```

> 2. The command below performs exporting to the trained model ``outputs/weights.pth`` in previous section and save exported model to the ``outputs/openvino/`` folder.

Reviewer: ```suggestion
2. The command below performs exporting of the trained model ``outputs/weights.pth`` in the previous section and saves exported model to the ``outputs/openvino/`` folder.
```

.. code-block::

  (detection) ...$ otx export otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                              --load-weights outputs/weights.pth
                              --save-model-to outputs/openvino/

  ...

  2022-12-29 01:39:11,980 | INFO : Exporting completed
  2022-12-29 01:39:11,980 | INFO : run task done.
  2022-12-29 01:39:11,990 | INFO : Exporting completed

> #TODO show how to run evaluation of exported model?

Reviewer: Yes. Let's show that accuracy didn't change and OV model is simple to validate

*************
Optimization
*************

1. ``otx optimize`` optimizes a model using NNCF or POT depending on the model format.

> - NNCF optimization used for trained snapshots in a framework-specific format such as checkpoint (pth) file from Pytorch. It starts training based on the weights from previous step in fewer precision.

Reviewer: ```suggestion
- NNCF optimization used for trained snapshots in a framework-specific format such as checkpoint (.pth) file from Pytorch. It starts training-aware quantization based on the obtained weights from the training stage.
```
Maybe that sounds better and gives some hint about NNCF algo.

> - POT optimization used for models exported in the OpenVINO IR format. It decreases the precision of the exported model and performs the post-training optimization.

Reviewer: ```suggestion
- POT optimization used for models exported in the OpenVINO™ IR format. It decreases floating-point precision to integer precision of the exported model by performing the post-training optimization.
```

  ...

  2022-12-29 02:11:49,018 | INFO : Loaded model weights from Task Environment
  2022-12-29 02:11:49,018 | INFO : Model architecture: ATSS
  2022-12-29 02:11:49,018 | INFO : Loaded model weights from Task Environment
  2022-12-29 02:11:49,018 | INFO : Model architecture: ATSS
  2022-12-29 02:11:49,019 | INFO : Task initialization completed
  INFO:nncf:Please, provide execution parameters for optimal model initialization
  2022-12-29 02:11:56,996 - mmdet - INFO - Received non-NNCF checkpoint to start training -- initialization of NNCF fields will be done
  [ INFO ] Received non-NNCF checkpoint to start training -- initialization of NNCF fields will be done
  [ INFO ]  Received non-NNCF checkpoint to start training -- initialization of NNCF fields will be done
  2022-12-29 02:11:56,999 - mmdet - INFO - Calculating an original model accuracy
  ...

  INFO:nncf:Original model accuracy: 0.4319
  INFO:nncf:Compressed model accuracy: 0.5564
  INFO:nncf:Absolute accuracy drop: -0.1245
  INFO:nncf:Relative accuracy drop: -28.82%
  INFO:nncf:Accuracy budget: 0.1345

#TODO significant drop of the loaded snapshot: 0.43 instead of 0.83

> #TODO show how to run evaluation of optimized model and its metrics?

Reviewer: After export we see the output metric. I think it is enough, but we can mention, that optimized models can be easily validated just like we did in the  previous sections

                                --train-ann-files BBCD/train/_annotations.coco.json 
                                --train-data-roots  BBCD/train 
                                --val-ann-files BBCD/valid/_annotations.coco.json 
                                --val-data-roots BBCD/valid 
                                --load-weights outputs/openvino/openvino.xml
                                --save-model-to outputs/pot

  ...

  2022-12-31 05:31:04,125 | INFO : Loading OpenVINO OTXDetectionTask
  2022-12-31 05:31:05,470 | INFO : OpenVINO task initialization completed
  2022-12-31 05:31:05,470 | INFO : Start POT optimization

  ...

  2022-12-31 05:37:51,004 | INFO : POT optimization completed
  2022-12-31 05:37:51,219 | INFO : Start OpenVINO inference
  2022-12-31 05:37:55,423 | INFO : OpenVINO inference completed
  2022-12-31 05:37:55,423 | INFO : Start OpenVINO metric evaluation
  2022-12-31 05:37:55,776 | INFO : OpenVINO metric evaluation completed
  Performance(score: 0.8343621399176954, dashboard: (1 metric groups))

The POT optimization will take 5-10 minutes without logging.

> The following stages how to deploy model and run demo are described in [link].

Reviewer: We already have a dedicated page for that, we may add the link now. The page is empty, but created

Author: Fixed

> The following stages how to deploy model and run demo are described in [link].

Reviewer: Also, I thought about adding some interaction with users. It just thoughts, but for now tutorial is too serious for me. :) We have strict sentences like: "Expample of command... Following stages: here... and so on. Maybe it is a good idea to water down just a little bit.
For example, add: "Now we have fully trained, optimized and exported in efficient model representation ready-to use object detection model. Following `link <>` provides further tutorial how to deploy and use your model in the demonstration mode."

Author: Yeah, it's a great idea, I like user-friendly tutorials as well :) I'll make it more kind

Author: Done

***************
Troubleshooting
***************

> #TODO possible error logs and their solution?

Reviewer: for example?

Author: I removed this block, because I don't have content for that. But let's keep in mind, that we can make such section, if needed

Dataset preparation
***************************

1. Clone a repository with `WGISD dataset <https://github.com/thsant/wgisd>`_. 

.. code-block::
  cd data
  git clone https://github.com/thsant/wgisd.git
  cd wgisd
  git checkout 6910edc5ae3aae8c20062941b1641821f0c30127

This dataset contains images of grapevines with the annotation for different varieties of grapes. 
- ``CDY``	- Chardonnay
- ``CFR``	- Cabernet Franc
- ``CSV``	- Cabernet Sauvignon
- ``SVB``	- Sauvignon Blanc
- ``SYH``	- Syrah

It's a great example to start with. The model achieves high accuracy right from the beginning of the training due to large and focused objects. Also, these objects are distinguished by a person, so we can check inference results just by looking at images.

> .. image:: ../../../../utils/images/wgisd_dataset_sample.jpg

Reviewer: Can we add the second image with bboxes for comparison?

Author: I didn't want to put 2 images here not to extend page too much. And  I wanted to use this image without annotation to run demo on.

  :width: 600
  :alt: this image uploaded from this `source <https://github.com/thsant/wgisd/blob/master/data/CDY_2015.jpg>`_

2. Check the file structure of downloaded repository, we will need the following files:

.. code-block::

  wgisd
  ├── README.md
  ├── coco_annotations/
      ├── train_bbox_instances.json
      └── test_bbox_instances.json
  ├──data/
      └── <images>
  ...

> 3. ``(Optional)`` To simplify the command line functions calling, we may create a ``data.yaml`` file with annotations info and pass it as a ``--data`` parameter. The content of the ``training_extesions/data.yaml`` for BCCD dataset should have absolute paths and will be similar to that:

Reviewer: wgisd

  {'data': 
    {'train': 
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'val':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/test_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'},
    'test':
      {'ann-files': '/home/<username>/training_extensions/data/wgisd/coco_annotations/train_bbox_instances.json',
       'data-roots': '/home/<username>/training_extensions/data/wgisd/data'}
    }
  }

> ``Ann-files`` contains a path to the annotation., while ``data-roots`` is a path to the folder, where images are stored.

Reviewer: ``Ann-files`` contains a path to the annotation**s,** while ``data-roots`` is a path to the folder, where images are stored.

These are needed as inputs for the further commands: ``export``, ``eval``,  ``optimize``,  ``deploy`` and ``demo``.

3. To have a specific example in this tutorial, all commands will be run on the ATSS model. It's a medium model, that achieves relatively high accuracy, while keeping the inference fast.

> The following command line starts 1 GPU training of the medium object detection model on BCCD dataset:

Reviewer: wgisd

5. The training results are ``weights.pth`` and ``label_schema.json`` files located in ``outputs`` folder, training logs can be found in the ``outputs/logs`` dir.

.. code-block::
  ...
  2023-01-10 05:40:21,520 | INFO : Update Lr patience: 3
  2023-01-10 05:40:21,520 | INFO : Update Validation Interval: 2
  2023-01-10 05:40:21,520 | INFO : Update Early-Stop patience: 5
  2023-01-10 05:40:23,140 | INFO : Epoch [1][1/31]        lr: 1.333e-03, eta: 11 days, 14:44:47, time: 1.619, data_time: 0.961, memory: 4673, current_iters: 0, loss_cls: 1.1261, loss_bbox: 0.6514, loss_centerness: 0.6337, loss: 2.4112, grad_norm: 18.5789

  ...
  2023-01-10 05:52:33,985 | INFO : run task done.
  2023-01-10 05:52:35,682 | INFO : Inference completed
  2023-01-10 05:52:35,683 | INFO : called evaluate()
  2023-01-10 05:52:35,907 | INFO : F-measure after evaluation: 0.5487693710118504
  2023-01-10 05:52:35,907 | INFO : Evaluation completed
  Performance(score: 0.5487693710118504, dashboard: (1 metric groups))

> After about 15 minutes of training we have the PyTorch object detection model trained with OTX, that we can use for evaliation, export, optimization and deployment. 

Reviewer: This time may very depending on the hardware. Let's avoid this. This is also related to POT time estimation

Author: I think, it's good to let user know, what to expect on this dataset. I agree, that it depends highly on the hardware, but I mentioned used hardware at the beginning of the page. 
Maybe, the compromise here will be to highlight that this characteristic depends on hardware, but keep this value for GeForce 3090.

Reviewer: Please, also add CPU info at the beginning of the page. Okay, agree
```suggestion
After about 15 minutes of training (depends on the hardware) we have the PyTorch object detection model trained with OTX, that we can use for evaliation, export, optimization and deployment. 
```

1. ``otx eval`` runs evaluation of a trained model on a particular dataset.

Eval function receives test annotation information and model snapshot, trained in previous step.

> Please note, that ``label schema.json`` file should be located in the same folder with model snaphot, as it contains meta information about the dataset .

Reviewer: ```suggestion
Please note, that ``label schema.json`` file should be located in the same folder with model snapshot, as it contains meta information about the dataset .
```

  2022-12-29 01:39:11,990 | INFO : Exporting completed

3. You can check the accuracy of exported model as simple as accuracy of the ``.pth`` model, using ``otx eval`` with the path of IR model.

.. code-block::

  (detection) ...$ otx eval otx/algorithms/detection/configs/detection/mobilenetv2_atss/template.yaml
                            --test-ann-files BBCD/valid/_annotations.coco.json 
                            --test-data-roots  BBCD/valid 
                            --load-weights outputs/openvino/openvino.xml
                            --save-performance outputs/openvino/performance.json
  
  ...

*************
Optimization
*************

1. ``otx optimize`` optimizes a model using NNCF or POT depending on the model format.

> - NNCF optimization is used for trained snapshots in a framework-specific format such as checkpoint (pth) file from Pytorch. It starts training-aware quantization based on the obtained weights from the training stage.

Reviewer: Sorry for misleading, according to the NNCF docs, the right naming is: Accuracy-Aware-training

```suggestion
- NNCF optimization is used for trained snapshots in a framework-specific format such as checkpoint (.pth) file from Pytorch. It starts Accuracy-Aware quantization based on the obtained weights from the training stage.
```
